cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3758 @@
+        {
+        public:
+            using OutputBatchCallback = std::function<void(ExecBatch)>;
+            using FinishedCallback = std::function<void(int64_t)>;
+            using GenerateFn = std::function<Status(size_t)>;
+            using ScheduleCallback = std::function<Status(GenerateFn)>;
+            using AbortCallback = std::function<void()>;
+
+            virtual Status Init(
+                std::vector<std::string> columns,
+                int scale_factor,
+                int64_t batch_size) = 0;
+
+            virtual Status StartProducing(
+                size_t num_threads,
+                OutputBatchCallback output_callback,
+                FinishedCallback finished_callback,
+                ScheduleCallback schedule_callback) = 0;
+
+            void Abort(AbortCallback abort_callback)
+            {
+                bool expected = false;
+                if(done_.compare_exchange_strong(expected, true))
+                {
+                    abort_callback();
+                }
+            }
+
+            virtual std::shared_ptr<Schema> schema() const = 0;
+
+            virtual ~TpchTableGenerator() = default;
+
+        protected:
+            std::atomic<bool> done_ = { false };
+            std::atomic<int64_t> batches_generated_ = { 0 };
+        };
+
+        int GetNumDigits(int64_t x)
+        {
+            // This if statement chain is for MAXIMUM SPEED
+            /*
+              .,
+              .      _,'f----.._
+              |\ ,-'"/  |     ,'
+              |,_  ,--.      /
+              /,-. ,'`.     (_
+              f  o|  o|__     "`-.
+              ,-._.,--'_ `.   _.,-`
+              `"' ___.,'` j,-'
+              `-.__.,--'
+             */
+            // Source: https://stackoverflow.com/questions/1068849/how-do-i-determine-the-number-of-digits-of-an-integer-in-c
+            ARROW_DCHECK(x >= 0);
+            if(x < 10ll) return 1;
+            if(x < 100ll) return 2;
+            if(x < 1000ll) return 3;
+            if(x < 10000ll) return 4;
+            if(x < 100000ll) return 5;
+            if(x < 1000000ll) return 6;
+            if(x < 10000000ll) return 7;
+            if(x < 100000000ll) return 8;
+            if(x < 1000000000ll) return 9;
+            if(x < 10000000000ll) return 10;
+            if(x < 100000000000ll) return 11;
+            if(x < 1000000000000ll) return 12;
+            if(x < 10000000000000ll) return 13;
+            if(x < 100000000000000ll) return 14;
+            if(x < 1000000000000000ll) return 15;
+            if(x < 10000000000000000ll) return 16;
+            if(x < 100000000000000000ll) return 17;
+            if(x < 1000000000000000000ll) return 18;
+            return -1;
+        }
+
+        void AppendNumberPaddedToNineDigits(char *out, int64_t x)
+        {
+            // We do all of this to avoid calling snprintf, which does a lot of crazy
+            // locale stuff. On Windows and MacOS this can get suuuuper slow
```

Comment:
We have formatting utilities in `arrow/util/formatting.h`, no need to reinvent them.

Reply:
@save-buffer looks like this comment is still unaddressed.  Also, if speed is a concern, I notice [~2x faster performance](https://github.com/westonpace/arrow/commit/d9efc98634ceabc09cf187c4d511174efb0eb272) with [`arrow::internal::detail::FormatAllDigitsLeftPadded`](https://github.com/apache/arrow/blob/91f65857c41d580dcf6db563bb2ca776b3eef0d7/cpp/src/arrow/util/formatting.h#L129) than with `AppendNumberPaddedToNineDigits`